lib/WasmReader/WasmSignature.cpp

```diff
@@ -141,24 +146,40 @@ void WasmSignature::FinalizeSignature()
         }
     }
 
+    // 3 bits for result type, 3 for each arg
+    const uint32 nBitsForResult = 3;
+#ifdef ENABLE_WASM_SIMD
+    const uint32 nBitsForArgs = 3;
+#else
+    // We can drop 1 bit by excluding void
+    const uint32 nBitsForArgs = 2;
+#endif
     CompileAssert(Local::Void == 0);
+    // Make sure we can encode all types (including void) with the number of bits reserved
+    CompileAssert(Local::Limit <= (1 << nBitsForResult));
+    // Make sure we can encode all types (excluding void) with the number of bits reserved
+    CompileAssert(Local::Limit - 1 <= (1 << nBitsForArgs));
+
+    ::Math::RecordOverflowPolicy sigOverflow;
+    const uint32 bitsRequiredForSig = UInt32Math::MulAdd<nBitsForArgs, nBitsForResult>((uint32)paramCount, sigOverflow);
+    if (sigOverflow.HasOverflowed())
+    {
+        return;
+    }
 
-#if 0
-    CompileAssert(Local::Limit - 1 <= 4);
-
-    // 3 bits for result type, 2 for each arg
     // we don't need to reserve a sentinel bit because there is no result type with value of 7
-    uint32 sigSize = ((uint32)paramCount) * 3 + 3;
-    if (sigSize <= sizeof(m_shortSig) << 3)
+    CompileAssert(Local::Limit <= 0b111);
```

Comment:
Wouldn't the first two `CompileAssert`s cover this last one  `CompileAssert`?

Reply:
It does, but they both convey 2 different things.The first is to make sure the enum does not go past the number of bits allocated.
The second is to make sure we don't have a enum value with only `1`s